### DISCUSSION TITLE
Please reconsider deprecating "name" in now.json

### DISCUSSION THREAD

**Discussion by simonw** [vercel/vercel#4021] [General]
Continuation of a conversation in https://github.com/simonw/datasette-publish-now/issues/3

I'm building a Datasette plugin for deploying SQLite databases to Now v2: https://github.com/simonw/datasette-publish-now

The desired workflow is this:

* You create a SQLite database - using one of [these tools](https://datasette.readthedocs.io/en/stable/ecosystem.html#tools-for-creating-sqlite-databases), or by running code in a Jupyter notebook, or by some other mechanism
* You want to share that SQLite database with the world (let's say you're a data journalist working for a news publication and you want to publish the data that underlies your story)
* So... you install Now CLI, Datasette and the datasette-publish-now plugin, and you run:
* `datasette publish now2 covid.db --project=covid-19-cases`
* ... and you're done! https://covid-19-cases.now.sh/ is now live

It looks like the intent of deprecating `"name"` is to encourage everyone to use project linking instead: https://zeit.co/docs/now-cli#commands/now/project-linking

The problem is... project linking is a concept that doesn't fit with my above workflow at all. I don't want my users to have to think about GitHub repos, projects or even source code on their local computer - I want them to say "publish this database on the internet" and not have to ever think about that project again.

So please, reconsider removing the `"name"` field! Or help me work out how my workflow and tooling can work in its absence.

Here's the code where I use the `"name"` field in my publish plugin: https://github.com/simonw/datasette-publish-now/blob/85e50fb099cd6298c8fdf68e483ede088e9711c6/datasette_publish_now/__init__.py#L107-L117

**Comment by simonw** [comment]
Relevant blog entry: https://simonwillison.net/2020/Apr/8/weeknotes-zeit-now-v2/

**Reply by styfle** [reply]
Great write up, thanks!

**Comment by simonw** [comment]
Could I fix this by dynamically generating a `.now` directory?

I looked in a generated one now and saw this:
```
$ cat .now/project.json 
{
  "projectId": "QmWC6N9oXEWrd1tffNDp75QMHTFTotTBJfM88jyhdyUkFc",
  "orgId": "dnzucqM5NdghFaFTQtP0xXDc"
}
```
But I'm not sure how I would generate that `projectId` and `orgId` programatically.

**Reply by styfle** [reply]
This probably won't work for your use case because the project id doesn't exist until the first deployment. I think the `--name` flag is probably better. After the first deployment, the `.now` folder will be created and it won't be necessary to assign the name any longer because the cwd will be linked to the ZEIT Project.

**Reply by Vadorequest** [reply]
@styfle Indeed, and that's actually "boring" that we cannot know our orgId nor projectId from the GUI.
Multiple times I've looked for it on the GUI because I manage multiple teams and it would have been great to access this information without deploying an app, when cloning an existing project on a new computer, for instance.

Long story short, there should be a way to access to orgId and projectId from Zeit.co website.

**Reply by styfle** [reply]
@Vadorequest Would a new command like `now link` that runs through the same prompts as the first deployment with `now` (but not deploy) solve your problem?

**Reply by Vadorequest** [reply]
I believe it would help, yes.

Would be useful to also to print that ".now/project.json" has been created and print its content upon success.

**Reply by styfle** [reply]
I forgot to follow up here: we added a `vercel link` command so you can generate this hidden file and link to an existing project or setup a new project 👍

**Reply by denniswenger** [reply]
Unless I'm missing something, I still don't see how we can use this to automate what project it should link to. 
If providing --confirm it seems to make an assumption about what project to link to. 
Is there something I don't understand here?

**Reply by Vadorequest** [reply]
@denniswenger Take a look at https://github.com/UnlyEd/next-right-now/tree/v2-mst-aptd-at-lcz-sty

When running `yarn deploy:all` locally, I have a script that removes the .vercel folder between builds to regenerate it every time and avoid the same project to be re-built (which is what happens if the .vercel is kept)

**Comment by leo** [comment]
Absolutely agreed! Now CLI is currently – in fact – lacking a way to automatically pre-fill the values that are used when the project is linked. So far, we've only added two environment variables that let you set the Project ID and Org ID directly, so that you can link to an existing project.

To complete the workflow, however, we need to also add one that, for example, let's you insert the name. At the moment, that is done via `--name` and the `name` property in `now.json`. The deprecation message should instead be a suggestion message and we'll make sure to add a environment variable that can achieve the same thing.

**Reply by Vadorequest** [reply]
@leo I understand better. I'll still need a name in the future because that's how I match a config file with a project in my CI/CD using github actions.

If it's removed, it'll break the whole thing.

If it's meant to be replaced by something else, it may still work. But please consider multi single-tenant deployments where the same base code can deploy multiple Zeit projects using a different config file, such as https://github.com/UnlyEd/next-right-now/tree/v1-ssr-mst-aptd-gcms-lcz-sty

**Reply by TrimHall** [reply]
Was this issue ever resolved? I'm working on a project with a requirement to deploy the same codebase to multiple projects. The `name` option is the obvious solution but it seems to still be deprecated - unless there's another workaround that has been released and I've not yet stumbled across it?

**Comment by TrimHall** [comment]
Please correct me if I'm wrong, but `vercel link` seems to be the command that is run by default if there's no `.vercel` directory already created?

The challenge I have is that I need to deploy from gitlab CI, to 2 different environments - staging and prod. The obvious way to do this would be to link vercel and our gitlab repository and have the staging branch deploy to our staging url and main branch deploy to the prod url. 

However, due to sensitive data in the gitlab repo I'm unable to use this approach, so instead have set up one project, `staging-app` that deploys to staging and `app` that deploys to prod.

Is there a way I can respond to the vercel link commands in the CI .yml file and pass a project name? Essentially to have the equivalent of running the staging CI job to execute   `yarn vercel --name staging-app --scope mySelectedScope --token somevalidtoken --prod`

and for the production job `yarn vercel --name production-app --scope mySelectedScope --token somevalidtoken --prod`

I've tried writing in `printf "y \n \n y \n" | yarn vercel --scope mySelectedScope -- token somevalidtoken --prod` but always get a failed with SIGNIT after the first prompt executes

Any help would be hugely appreciated!

**Reply by Vadorequest** [reply]
Honestly, we're relying a lot on the `name` property, even though it's deprecated. Removing this property would break all our CI and everything would fall appart.

We also use a monorepo design where one repo is used to deploy multiple Vercel projects, and using the name property is how we select which customer gets deployed.

I surely hope it doesn't get removed.
